File: src/main/java/com/enderryno/nuclearcraft/abstracts/GenericCommand.java
package com.enderryno.nuclearcraft.abstracts;

import com.enderryno.nuclearcraft.annotations.CommandInfo;
import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public abstract class GenericCommand implements CommandExecutor {
    private final CommandInfo commandInfo;

    public GenericCommand() {
        commandInfo = getClass().getDeclaredAnnotation(CommandInfo.class);
        Objects.requireNonNull(commandInfo, "Commands must have a CommandInfo annotation");
    }

    public CommandInfo getCommandInfo() {
        return commandInfo;
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, @NotNull String[] args) {
        if (!getCommandInfo().permission().isEmpty()) {
            if (!sender.hasPermission(commandInfo.permission())) {
                sender.sendMessage(ChatColor.RED + "You can't execute this command.");
                return true;
            }
        }

        if (commandInfo.requiresPlayer()) {
            if(!(sender instanceof  Player)) {
                sender.sendMessage(ChatColor.RED + "This command can be only executed from a player.");
                return true;
            }
            execute((Player)sender, args);
            return true;

        }

        execute(sender, args);
        return true;
    }

    public void execute(Player player, String[] args) {}
    public void execute(CommandSender sender, String[] args) {}

}

File: src/main/java/com/mochibit/nuclearcraft/abstracts/GenericCommand.kt
package com.mochibit.nuclearcraft.abstracts

import com.mochibit.nuclearcraft.annotations.CommandInfo
import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import java.util.*

abstract class GenericCommand : CommandExecutor {
    val commandInfo: CommandInfo = javaClass.getDeclaredAnnotation(CommandInfo::class.java)

    init {
        Objects.requireNonNull(commandInfo, "Commands must have a CommandInfo annotation")
    }

    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        if (commandInfo.permission.isNotEmpty()) {
            if (!sender.hasPermission(commandInfo.permission)) {
                sender.sendMessage(ChatColor.RED.toString() + "You can't execute this command.")
                return true
            }
        }
        if (commandInfo.requiresPlayer) {
            if (sender !is Player) {
                sender.sendMessage(ChatColor.RED.toString() + "This command can be only executed from a player.")
                return true
            }
            execute(sender, args)
            return true
        }
        execute(sender, args)
        return true
    }

    open fun execute(player: Player, args: Array<String>) {}
    private fun execute(sender: CommandSender?, args: Array<String>?) {}
}
